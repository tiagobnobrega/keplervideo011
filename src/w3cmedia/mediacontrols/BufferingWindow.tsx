/*
 * Copyright 2022-2023 Amazon.com, Inc. or its affiliates. All rights reserved.
 *
 * AMAZON PROPRIETARY/CONFIDENTIAL
 *
 * You may not use this file except in compliance with the terms and
 * conditions set forth in the accompanying LICENSE.TXT file.
 *
 * THESE MATERIALS ARE PROVIDED ON AN "AS IS" BASIS. AMAZON SPECIFICALLY
 * DISCLAIMS, WITH RESPECT TO THESE MATERIALS, ALL WARRANTIES, EXPRESS,
 * IMPLIED, OR STATUTORY, INCLUDING THE IMPLIED WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE, AND NON-INFRINGEMENT.
 */

import LottieView from '@amzn/lottie-react-native';
import React from 'react';
import { StyleSheet, View, ViewProps } from 'react-native';
import { COLORS } from '../../styles/Colors';
import { scaleUxToDp } from '../../utils/pixelUtils';

const BufferingWindow = React.memo((props: ViewProps) => {
  return (
    <View {...props} style={styles.container} testID="buffering-view">
      <LottieView
        testID="video-buffer-view"
        source={require('../../js/resources/loader.json')}
        autoPlay
        loop
        style={styles.loader}
      />
    </View>
  );
});

export const styles = StyleSheet.create({
  container: {
    width: '100%',
    height: '100%',
    top: 0,
    left: 0,
    position: 'absolute',
    flex: 1,
    backgroundColor: COLORS.BLACK,
    flexDirection: 'column',
    alignItems: 'center',
    justifyContent: 'center',
    zIndex: 5,
  },
  loader: {
    width: scaleUxToDp(80),
    height: scaleUxToDp(320),
    alignSelf: 'center',
  },
});
export default BufferingWindow;
