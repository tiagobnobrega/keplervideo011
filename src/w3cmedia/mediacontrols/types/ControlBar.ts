import { VideoPlayer } from '@amzn/react-native-w3cmedia';

export enum PlayerControlType {
  PLAY = 'play',
  PAUSE = 'pause',
  PLAYPAUSE = 'playpause',
  SKIPBACKWARD = 'skip_backward',
  SKIPFORWARD = 'skip_forward',
}
export interface ControlBarButtonProps {
  videoRef: React.MutableRefObject<VideoPlayer | null>;
  onBlur?: () => void;
  playerControlType?: string;
}
