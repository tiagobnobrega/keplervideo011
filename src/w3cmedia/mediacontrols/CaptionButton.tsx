import { TVFocusGuideView } from '@amzn/react-native-kepler';
import React, { useEffect, useRef } from 'react';
import { StyleSheet, TouchableOpacity } from 'react-native';
import { DPADEventType } from '../../constants';
import { COLORS } from '../../styles/Colors';
import { scaleUxToDp } from '../../utils/pixelUtils';
import PlayerButton from './PlayerButton';
import { CaptionButtonProps } from './types/Captions';

const CaptionEnabledIcon = 'closed-caption';
const CaptionDisabledIcon = 'closed-caption-disabled';
const CaptionUnavailableIcon = 'closed-caption-off';

const CaptionButton = React.memo(
  ({
    onPress,
    video,
    captionMenuVisibility,
    testID,
    playerControlType,
    isCaptionButtonFocused,
  }: CaptionButtonProps) => {
    const captionsExist = video?.textTracks?.length > 0 ? true : false;
    const captionButtonRef = useRef<TouchableOpacity>(null);
    const onBlurCaptionButton = () => {
      captionButtonRef?.current?.blur();
    };
    let CaptionIcon = CaptionUnavailableIcon;
    if (captionsExist) {
      CaptionIcon = CaptionEnabledIcon;
    } else {
      CaptionIcon = CaptionDisabledIcon;
    }
    useEffect(() => {
      if (
        playerControlType === DPADEventType.BACK &&
        !captionMenuVisibility &&
        isCaptionButtonFocused
      ) {
        captionButtonRef.current?.focus();
      }
    }, [playerControlType, captionMenuVisibility, isCaptionButtonFocused]);

    return (
      <TVFocusGuideView trapFocusLeft={captionMenuVisibility}>
        <PlayerButton
          ref={captionButtonRef}
          onBlur={onBlurCaptionButton}
          onPress={onPress}
          icon={CaptionIcon}
          size={40}
          overrideStyle={
            captionMenuVisibility ? styles.captionSelected : undefined
          }
          testID={testID}
        />
      </TVFocusGuideView>
    );
  },
);

const styles = StyleSheet.create({
  captionSelected: {
    backgroundColor: COLORS.DARK_GRAY + 'D9',
    borderRadius: scaleUxToDp(40),
  },
});
export default CaptionButton;
