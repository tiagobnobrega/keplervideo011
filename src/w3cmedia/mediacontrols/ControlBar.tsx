import { VideoPlayer } from '@amzn/react-native-w3cmedia';
import React from 'react';
import { StyleSheet, View } from 'react-native';
import CaptionButton from './CaptionButton';

export interface ControlBarProps {
  captions: () => void;
  videoRef: React.MutableRefObject<VideoPlayer | null>;
  captionMenuVisibility: boolean;
  playerControlType?: string;
  isCaptionButtonFocused: boolean | null;
}
export const ControlBar = ({
  captions,
  videoRef,
  captionMenuVisibility,
  playerControlType,
  isCaptionButtonFocused,
}: ControlBarProps) => {
  return (
    <View style={styles.controlBar} testID="control-bar">
      <CaptionButton
        playerControlType={playerControlType}
        video={videoRef.current as VideoPlayer}
        onPress={captions}
        testID={'video-player-caption-btn'}
        captionMenuVisibility={captionMenuVisibility}
        isCaptionButtonFocused={isCaptionButtonFocused}
      />
    </View>
  );
};

export const styles = StyleSheet.create({
  controlBar: {
    flexDirection: 'row',
    justifyContent: 'flex-end',
    alignItems: 'center',
  },
});
