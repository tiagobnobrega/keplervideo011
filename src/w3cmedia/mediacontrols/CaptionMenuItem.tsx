import React from 'react';
import { StyleSheet, Text } from 'react-native';
import FocusableElement from '../../components/FocusableElement';
import { COLORS } from '../../styles/Colors';
import { scaleUxToDp } from '../../utils/pixelUtils';
import { CaptionMenuItemProps } from './types/Captions';

const CaptionMenuItem = React.memo(
  ({ text, selected, onPress, testID }: CaptionMenuItemProps) => {
    return (
      <FocusableElement
        onPress={onPress}
        style={selected && styles.selectedMenuItem}
        onFocusOverrideStyle={styles.focusedMenuItem}
        testID={testID}>
        <Text style={styles.menuItemText} testID={'video-player-caption-menu'}>
          {text}
        </Text>
      </FocusableElement>
    );
  },
);
export default CaptionMenuItem;

const styles = StyleSheet.create({
  menuItemText: {
    color: COLORS.WHITE,
    fontWeight: '500',
    fontSize: scaleUxToDp(30),
    marginVertical: scaleUxToDp(4),
    padding: scaleUxToDp(20),
  },
  focusedMenuItem: { backgroundColor: COLORS.GRAY + 'E6' },
  selectedMenuItem: { backgroundColor: COLORS.GRAY + '66' },
});
