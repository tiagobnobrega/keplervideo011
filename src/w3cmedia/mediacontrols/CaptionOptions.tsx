import React from 'react';
import CaptionMenuItem from './CaptionMenuItem';
import { CaptionOptionProps } from './types/Captions';

const CaptionOptions = React.memo(
  ({ enableCaption, selectedCaptionId, video }: CaptionOptionProps) => {
    const Options: React.JSX.Element[] = [];
    const textTrackList = video?.textTracks || [];
    for (
      let captionIndex = 0;
      captionIndex < textTrackList.length;
      captionIndex++
    ) {
      const textTrack = textTrackList[captionIndex];
      if (textTrack) {
        const { label, id } = textTrack;
        Options.push(
          <CaptionMenuItem
            key={label}
            text={label}
            index={id}
            selected={id === selectedCaptionId}
            onPress={() => {
              enableCaption(id);
            }}
          />,
        );
      }
    }
    return <>{Options}</>;
  },
);

export default CaptionOptions;
