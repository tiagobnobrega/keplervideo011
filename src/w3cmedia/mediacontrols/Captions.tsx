import { TVFocusGuideView } from '@amzn/react-native-kepler';
import React, { useCallback, useState } from 'react';
import { StyleSheet, View } from 'react-native';
import { CAPTION_DISABLE_ID, CAPTION_DISABLE_TEXT } from '../../constants';
import { COLORS } from '../../styles/Colors';
import { scaleUxToDp } from '../../utils/pixelUtils';
import CaptionMenuItem from './CaptionMenuItem';
import CaptionOptions from './CaptionOptions';
import { CaptionMenuProps } from './types/Captions';

export const CaptionMenu = ({
  captionMenuVisibility,
  video,
  setSelectedCaption,
}: CaptionMenuProps) => {
  const [captionID, setCaptionID] = useState<string>(CAPTION_DISABLE_ID);

  const removeCurrentCaption = useCallback(() => {
    if (captionID !== CAPTION_DISABLE_ID) {
      console.log('Disabling Captions');
      if (video && video.textTracks) {
        video.textTracks.getTrackById(captionID)!.mode = 'hidden';
      }
    }
    setCaptionID(CAPTION_DISABLE_ID);
  }, [captionID, video]);

  const enableCaption = useCallback(
    (id: string) => {
      if (id === captionID) {
        return;
      }
      removeCurrentCaption();
      if (video && video.textTracks) {
        video.textTracks.getTrackById(id)!.mode = 'showing';
        setCaptionID(id);
        setSelectedCaption(id);
      }
      setCaptionID(id);
      setSelectedCaption(id);
    },
    [captionID, removeCurrentCaption, video, setSelectedCaption],
  );

  const turnOffCaptions = useCallback(() => {
    setSelectedCaption(CAPTION_DISABLE_ID);
    removeCurrentCaption();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return (
    <TVFocusGuideView trapFocusLeft trapFocusRight trapFocusUp>
      {captionMenuVisibility ? (
        <View style={styles.menu}>
          <CaptionOptions
            video={video}
            enableCaption={enableCaption}
            selectedCaptionId={captionID}
          />
          <CaptionMenuItem
            key={CAPTION_DISABLE_TEXT}
            text={CAPTION_DISABLE_TEXT}
            onPress={turnOffCaptions}
            selected={captionID === CAPTION_DISABLE_ID}
            testID={'caption-menu-focus-element'}
          />
        </View>
      ) : null}
    </TVFocusGuideView>
  );
};

const styles = StyleSheet.create({
  menu: {
    backgroundColor: COLORS.BLACK + 'E6',
    width: scaleUxToDp(250),
    justifyContent: 'center',
    zIndex: 30,
    borderRadius: scaleUxToDp(15),
    overflow: 'hidden',
  },
  captionSelected: {
    backgroundColor: COLORS.DARK_GRAY + 'D9',
    borderRadius: scaleUxToDp(40),
  },
});
