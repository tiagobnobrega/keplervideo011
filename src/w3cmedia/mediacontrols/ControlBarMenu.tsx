import { VideoPlayer } from '@amzn/react-native-w3cmedia';
import React from 'react';
import { StyleSheet, View } from 'react-native';
import { scaleUxToDp } from '../../utils/pixelUtils';
import { CaptionMenu } from './Captions';

interface ControlBarMenuProps {
  captionMenuVisibility: boolean;
  videoRef: React.MutableRefObject<VideoPlayer | null>;
  setSelectedCaptionInMenuBar: (id: string) => void;
}
const ControlBarMenu = React.memo(
  ({
    captionMenuVisibility,
    videoRef,
    setSelectedCaptionInMenuBar,
  }: ControlBarMenuProps) => {
    const setSelectedCaption = (id: string) => {
      setSelectedCaptionInMenuBar(id);
    };

    return (
      <View style={styles.playerMenusContainer}>
        <CaptionMenu
          captionMenuVisibility={captionMenuVisibility}
          video={videoRef.current as VideoPlayer}
          setSelectedCaption={setSelectedCaption}
        />
      </View>
    );
  },
);

const styles = StyleSheet.create({
  playerMenusContainer: {
    position: 'absolute',
    bottom: scaleUxToDp(90),
    right: 0,
  },
});

export default ControlBarMenu;
