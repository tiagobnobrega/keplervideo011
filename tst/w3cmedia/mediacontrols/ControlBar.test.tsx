/*
 * Copyright (c) 2024 Amazon.com, Inc. or its affiliates.  All rights reserved.
 *
 * PROPRIETARY/CONFIDENTIAL.  USE IS SUBJECT TO LICENSE TERMS.
 */
import { VideoPlayer } from '@amzn/react-native-w3cmedia';
import { describe } from '@jest/globals';
import '@testing-library/jest-native/extend-expect';
import { render } from '@testing-library/react-native';
import React from 'react';
import { ControlBar } from '../../../src/w3cmedia/mediacontrols/ControlBar';

const videoRef: React.MutableRefObject<VideoPlayer | null> = {
  current: {
    textTracks: {
      length: 2,
    },
  } as unknown as VideoPlayer,
};
const renderCaptionButton = () => {
  return (
    <ControlBar
      captions={jest.fn()}
      videoRef={videoRef}
      captionMenuVisibility={false}
      isCaptionButtonFocused={true}
    />
  );
};
describe('ControlBarte sts', () => {
  it('component renders correctly', () => {
    const component = render(renderCaptionButton());
    expect(component).toMatchSnapshot();
  });
});
