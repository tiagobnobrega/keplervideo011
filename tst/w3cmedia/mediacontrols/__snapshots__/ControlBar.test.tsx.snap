// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ControlBarte sts component renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flexDirection": "row",
      "justifyContent": "flex-end",
    }
  }
  testID="control-bar"
/>
`;
